File: src/cube.cpp
#include <algorithm>
#include <cstddef>
#include <cstdint>
#include <cstdio>
#include <cstdlib>
#include <filesystem>
#include <string>
#include <vector>

#include "corner_heuristic.hpp"
#include "corner_orientation.hpp"
#include "corner_position.hpp"
#include "cube.hpp"
#include "cube_bridge.hpp"
#include "edge_heuristic.hpp"
#include "edge_orientation.hpp"
#include "edge_position.hpp"
#include "logger.hpp"
#include "utils.hpp"
#include "settings.hpp"


std::vector<uint16_t> Cube::corner_orientations;
std::vector<uint16_t> Cube::corner_positions;
std::vector<uint16_t> Cube::corner_heuristics;

std::vector<uint16_t> Cube::edge_orientations;
std::vector<uint32_t> Cube::edge_positions;
std::vector<uint8_t> Cube::edge_heuristics;


uint8_t GetRevRotation(uint8_t rotation) {
    if (rotation % 2 == 0) {
        return rotation + 1;
    }
    return rotation - 1;
}


void Cube::Initialize() {
    if (!std::filesystem::exists(GetFilePath(""))) {
        if (std::filesystem::create_directories(GetFilePath(""))) {
            LOG_ALL("Create precomputation folder for binaries");
        }
        else {
            LOG_ERROR("Failed to create folder for binaries");
        }
    }

    LoadOrGenerate("corner_orientations.bin", corner_orientations, kCornerOrientationSize,
        [&](){CornerOrientationInitialization(corner_orientations);}, "[1/6] Corner Orientations");

    LoadOrGenerate("corner_positions.bin", corner_positions, kCornerPositionsSize,
        [&](){CornerPositionInitialization(corner_positions);}, "[2/6] Corner Positions");

    LoadOrGenerate("corner_heuristics.bin", corner_heuristics, kNumCornerHeuristic,
        [&](){CornerHeuristicInitialization(corner_orientations, corner_positions, corner_heuristics);}, "[3/6] Corner Heuristics");

    LoadOrGenerate("edge_orientations.bin", edge_orientations, kEdgeOrientationSize,
        [&](){EdgeOrientationInitialization(edge_orientations);}, "[4/6] Edge Orientations");

    LoadOrGenerate("edge_positions.bin", edge_positions, kEdgePositionsSize,
        [&](){EdgePositionInitialization(edge_positions);}, "[5/6] Edge Positions");

    LoadOrGenerate("edge_heuristics.bin", edge_heuristics, kNumEdgeHeuristic,
        [&](){EdgeHeuristicInitialization(edge_orientations, edge_positions, 0, 0, edge_heuristics);}, "[6/6] Edge Heuristics");
}


constexpr std::array<uint8_t, kNumRotations> kLegalMoveIndex = {
    8, 9, 8, 9, 10, 11, 10, 11, 12, 13, 12, 13, 0, 0, 0, 0, 0, 0
};


std::pair<bool, State> Cube::Rotate(const State& prev_state, const uint8_t& rotation) {
    uint16_t corner_orientation = prev_state.hash_2 >> 20;      // 12 bites         NOLINT
    uint16_t corner_position = prev_state.hash_1;               // 16 bites         NOLINT
    uint16_t edge_orientation = prev_state.hash_3 >> 20;        // 11 bites         NOLINT
    uint32_t edge_position_1 = prev_state.hash_2 & ((1<<20)-1); // 20 bites         NOLINT
    uint32_t edge_position_2 = prev_state.hash_3 & ((1<<20)-1); // 20 bites         NOLINT
    if (kLegalMoveIndex[rotation] != 0 && ((corner_heuristics[(corner_orientation*kNumCornerPositions) + corner_position] >> kLegalMoveIndex[rotation]) & 1) == 0) {
        return {false, prev_state};
    }
    corner_orientation = corner_orientations[(corner_orientation*kNumRotations) + rotation];
    corner_position = corner_positions[(corner_position*kNumRotations) + rotation];
    edge_orientation = edge_orientations[(edge_orientation*kNumRotations) + rotation];
    edge_position_1 = edge_positions[(edge_position_1*kNumRotations) + rotation];
    edge_position_2 = edge_positions[(edge_position_2*kNumRotations) + rotation];
    return {true, State(corner_orientation, corner_position, edge_orientation, edge_position_1, edge_position_2)};
}


void Cube::UploadComputationToDevice() {
    #ifdef USE_CUDA
    if (Settings::UseCuda()) {
        LOG_EXTRA("Start Cube Uploading Precomputation to Device");
        UploadCubeComputationToDevices(corner_orientations, corner_positions, corner_heuristics, edge_orientations, edge_positions, edge_heuristics);
        LOG_INFO("Cube Precomputation Uploaded to Device");
        LOG_MEMORY();
    }
    #endif // USE_CUDA
}


uint16_t Cube::GetCurCornerHeuristic(const State& state) {
    uint16_t corner_orientation = state.hash_2 >> 20;      // 12 bites         NOLINT
    uint16_t corner_position = state.hash_1;               // 16 bites         NOLINT
    return uint8_t(corner_heuristics[(corner_orientation*kNumCornerPositions) + corner_position] & ((uint16_t(1) << 8) - 1)); // NOLINT
}


void Cube::SetCurCornerHeuristic(const State& state) {
    cur_corner_heuristic_ = GetCurCornerHeuristic(state);
}

void Cube::SetCurEdgeHeuristic1(const State& state) {
    uint32_t orientation = state.hash_3 >> 20; // NOLINT
    uint32_t position = state.hash_2 & ((uint32_t(1) << 20) - 1); // NOLINT
    cur_edge_heuristic_1_ = edge_heuristics[(orientation*kNumEdgePositions) + position];
}

void Cube::SetCurEdgeHeuristic2(const State& state) {
    uint32_t orientation = state.hash_3 >> 20; // NOLINT
    orientation |= (std::popcount(orientation)%2) << (kNumEdges-1); // get last bit using even num bits parity
    uint32_t orientation_r = 0;
    for (int i = 1; i < kNumEdges; i++) {
        orientation_r |= ((orientation >> i) & uint32_t(1)) << (kNumEdges-1-i);
    }

    uint32_t position = state.hash_3 & ((uint32_t(1) << 20) - 1); // NOLINT
    uint32_t position_r = 0;
    uint32_t temp = kNumEdgePositions;
    for (int i = kNumEdges-1; i >= 6; i--) { // NOLINT
        temp /= i+1;
        position_r *= i+1;
        position_r += i - ((position / temp) % (i + 1)); // NOLINT
    }
    cur_edge_heuristic_2_ = edge_heuristics[(orientation_r*kNumEdgePositions) + position_r];
}


uint8_t Cube::GetMaxHeuristic(const State& state) {
    if (cur_corner_heuristic_ == uint8_t(-1)) {
        SetCurCornerHeuristic(state);
    }
    if (cur_edge_heuristic_1_ == uint8_t(-1)) {
        SetCurEdgeHeuristic1(state);
    }
    if (cur_edge_heuristic_2_ == uint8_t(-1)) {
        SetCurEdgeHeuristic2(state);
    }
    return std::max({cur_corner_heuristic_, cur_edge_heuristic_1_, cur_edge_heuristic_2_});
}


float Cube::GetAppHeuristic(const State& state) {
    if (cur_corner_heuristic_ == uint8_t(-1)) {
        SetCurCornerHeuristic(state);
    }
    if (cur_edge_heuristic_1_ == uint8_t(-1)) {
        SetCurEdgeHeuristic1(state);
    }
    if (cur_edge_heuristic_2_ == uint8_t(-1)) {
        SetCurEdgeHeuristic2(state);
    }
    return cur_corner_heuristic_ + cur_edge_heuristic_1_ + cur_edge_heuristic_2_;
}

File: analyser/cube.py
import numpy as np


def GetRotationFromStr(my_str):
    conversion = {"R": 0 ,
                  "R'": 1 ,
                  "L": 2 ,
                  "L'": 3 ,
                  "U": 4 ,
                  "U'": 5 ,
                  "D": 6 ,
                  "D'": 7 ,
                  "F": 8 ,
                  "F'": 9 ,
                  "B": 10,
                  "B'": 11,
                  "M": 12,
                  "M'": 13,
                  "E": 14,
                  "E'": 15,
                  "S": 16,
                  "S'": 17}
    return conversion[my_str]


def GetRevRotation(rotation):
    if rotation % 2 == 0:
        return rotation + 1
    return rotation - 1


class Cube:
    corner_orientations = []
    corner_positions = []
    corner_heuristics = []

    edge_orientations = []
    edge_positions = []
    edge_heuristics = []

    kNumEdges = 12
    kNumRotations = 18
    kNumCornerPositions = 40320
    kNumEdgePositions = 665280
    kLegalMoveIndex = [8, 9, 8, 9, 10, 11, 10, 11, 12, 13, 12, 13, 0, 0, 0, 0, 0, 0]


    @staticmethod
    def Initatlization():
        Cube.corner_orientations = np.fromfile("../precomputation/corner_orientations.bin", dtype=np.uint16)
        Cube.corner_positions = np.fromfile("../precomputation/corner_positions.bin", dtype=np.uint16)
        Cube.corner_heuristics = np.fromfile("../precomputation/corner_heuristics.bin", dtype=np.uint16)

        Cube.edge_orientations = np.fromfile("../precomputation/edge_orientations.bin", dtype=np.uint16)
        Cube.edge_positions = np.fromfile("../precomputation/edge_positions.bin", dtype=np.uint32)
        Cube.edge_heuristics = np.fromfile("../precomputation/edge_heuristics.bin", dtype=np.uint8)


    def __init__(self):
        self.corner_orientation = 0
        self.corner_position = 0

        self.edge_orientation = 0
        self.edge_position_1 = 0
        self.edge_position_2 = 665279

        self.GetHeuristic()


    def GetHeuristic(self):
        self.heuristic_corner = int(Cube.corner_heuristics[self.corner_orientation*Cube.kNumCornerPositions + self.corner_position] & ((1 << 8) - 1))
        self.heuristic_edge_1 = int(Cube.edge_heuristics[(self.edge_orientation*Cube.kNumEdgePositions) + self.edge_position_1])

        orientation = self.edge_orientation | ((self.edge_orientation.bit_count()%2) << (Cube.kNumEdges-1))
        orientation_r = 0
        for i in range(1, Cube.kNumEdges):
            orientation_r = orientation_r | ((orientation >> i) & 1) << (Cube.kNumEdges-1-i)
        position = self.edge_position_2
        position_r = 0
        temp = Cube.kNumEdgePositions
        for i in range(Cube.kNumEdges-1, 6-1, -1):
            temp = temp // (i+1)
            position_r *= i+1
            position_r += i - ((position // temp) % (i + 1))
        self.heuristic_edge_2 = int(Cube.edge_heuristics[(orientation_r*Cube.kNumEdgePositions) + position_r])


    def Rotate(self, rotation):
        if Cube.kLegalMoveIndex[rotation] != 0 and ((Cube.corner_heuristics[(self.corner_orientation*Cube.kNumCornerPositions) + self.corner_position] >> Cube.kLegalMoveIndex[rotation]) & 1) == 0:
            print("ILLEGAL")
            return False

        self.corner_orientation = int(Cube.corner_orientations[(self.corner_orientation*Cube.kNumRotations) + rotation])
        self.corner_position = int(Cube.corner_positions[(self.corner_position*Cube.kNumRotations) + rotation])

        self.edge_orientation = int(Cube.edge_orientations[(self.edge_orientation*Cube.kNumRotations) + rotation])
        self.edge_position_1 = int(Cube.edge_positions[(self.edge_position_1*Cube.kNumRotations) + rotation])
        self.edge_position_2 = int(Cube.edge_positions[(self.edge_position_2*Cube.kNumRotations) + rotation])

        self.GetHeuristic()


if __name__ == "__main__":
    Cube.Initatlization()
    cube = Cube()
    print(cube.heuristic_corner, cube.heuristic_edge_1, cube.heuristic_edge_2)
    cube.Rotate(0)
    print(cube.heuristic_corner, cube.heuristic_edge_1, cube.heuristic_edge_2)
